### PR TITLE
Detect if there's no network before attempting to make requests

### PR DESCRIPTION
**Pull request by AesteroidBlues**
Allowing libHttpClient to perform platform-dependent network detection before attempting to make requests. 

These platform calls ensure we are connected to some kind of network, regardless of whether or not that network is connected to the internet. Therefore, we should still retry as expected in transient network failures.

### CODE REVIEW

File: Source/HTTP/Android/android_http_request.cpp [hunk 1]
```diff
@@ -4,10 +4,11 @@
 #include <httpClient/httpClient.h>
 #include <vector>
 
-HttpRequest::HttpRequest(AsyncBlock* asyncBlock, JavaVM* javaVm, jclass httpRequestClass, jclass httpResponseClass) :
+HttpRequest::HttpRequest(AsyncBlock* asyncBlock, JavaVM* javaVm, jobject applicationContext, jclass httpRequestClass, jclass httpResponseClass) :
     m_httpRequestInstance(nullptr), 
     m_asyncBlock(asyncBlock),
     m_javaVm(javaVm),
+	m_applicationContext(applicationContext),
```
Comment (lubeltra) [inline Source/HTTP/Android/android_http_request.cpp]: tabs

File: Include/httpClient/httpClient.h [hunk 2]
```diff
@@ -702,3 +705,17 @@ hc_websocket_handle_t HCWebSocketDuplicateHandle(
...
+/// Used to wrap the JavaVM and ApplicationContext on Android devices.
+/// </summary>
+#if HC_PLATFORM == HC_PLATFORM_ANDROID
+typedef struct _HC_PLATFORM_INIT_ARGS {
+	JavaVM *JavaVM;
+	jobject ApplicationContext;
+} HC_PLATFORM_INIT_ARGS;
+#else 
+typedef struct _HC_PLATFORM_INIT_ARGS {
+	void* dummy;
```
Comment (lubeltra) [inline Include/httpClient/httpClient.h]: tabs everywhere

File: Source/HTTP/httpcall.cpp [hunk 3]
```diff
@@ -193,6 +193,11 @@ bool http_call_should_retry(
         return false;
     }
 
+	if (call->networkErrorCode == E_HC_NO_NETWORK)
+	{
+		return false;
+	}
```
Comment (matt-vanderkolk) [inline Source/HTTP/httpcall.cpp]: tabs

File: Build/libHttpClient.141.Android.Java/app/src/main/java/com/xbox/httpClient/HttpClientRequest.java [hunk 4]
```diff
@@ -20,6 +23,12 @@ public HttpClientRequest() {
         requestBuilder = new Request.Builder();
     }
 
+    public static boolean isNetworkAvailable(Context context) {
+        ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
+        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
```
Comment (matt-vanderkolk) [inline Build/libHttpClient.141.Android.Java/app/src/main/java/com/xbox/httpClient/HttpClientRequest.java]: Are there any concerns of platform support here on older API levels?

File: Include/httpClient/httpClient.h [hunk 5]
```diff
@@ -702,3 +705,17 @@ hc_websocket_handle_t HCWebSocketDuplicateHandle(
 STDAPI HCWebSocketCloseHandle(
     _In_ hc_websocket_handle_t websocket
     ) HC_NOEXCEPT;
+
+/// <summary>
+/// Used to wrap the JavaVM and ApplicationContext on Android devices.
+/// </summary>
+#if HC_PLATFORM == HC_PLATFORM_ANDROID
+typedef struct _HC_PLATFORM_INIT_ARGS {
```
Comment (matt-vanderkolk) [inline Include/httpClient/httpClient.h]: >_HC_PLATFORM_INIT_ARGS [](start = 15, length = 22)

Are we using this anywhere?

Comment (matt-vanderkolk) [reply Include/httpClient/httpClient.h]: nevermind

---
In reply to: [206337845](https://github.com/Microsoft/libHttpClient/pull/217#discussion_r206337845) [](ancestors = 206337845)

File: Source/HTTP/Android/android_http_request.h [hunk 6]
```diff
@@ -27,6 +27,7 @@ class HttpRequest {
     AsyncBlock* m_asyncBlock;
 
     JavaVM* m_javaVm;
+	jobject m_applicationContext;
```
Comment (matt-vanderkolk) [inline Source/HTTP/Android/android_http_request.h]: tabs

File: Build/libHttpClient.141.Android.Java/app/src/main/java/com/xbox/httpClient/HttpClientRequest.java (hunk 4 above)
Comment (AesteroidBlues) [reply Build/libHttpClient.141.Android.Java/app/src/main/java/com/xbox/httpClient/HttpClientRequest.java]: Nope this API has existed since level 1

**Pull request merged**